keps/sig-api-machinery/4192-svm-in-tree/README.md

```diff
@@ -0,0 +1,435 @@
+
+#### [Undecided] Changes while we move above APIs in-tree:
+To avoid any conflicts with the Storage Version Migrators running out of tree, we will change the _`group`_ from `migration.k8s.io` to `storagemigration.k8s.io`.
+
+The final APIs that will be moved in-tree are:
+- `v1alpha1` of `storageversionmigrations.storagemigration.k8s.io`
+
+### [Controller](https://github.com/kubernetes-sigs/kube-storage-version-migrator/tree/60dee538334c2366994c2323c0db5db8ab4d2838/pkg/controller) to move
+#### [Migrator Controller](https://github.com/kubernetes-sigs/kube-storage-version-migrator/tree/60dee538334c2366994c2323c0db5db8ab4d2838/pkg/migrator)
+Currently, the Storage Version Migrator comprises two controllers: the `Trigger` controller and the `Migrator` controller. The Trigger controller performs resource discovery, identifying supported resources with the preferred server version every `10 minutes`. Subsequently, the Trigger controller creates the `StorageVersionMigration` resource to initiate the migration process. The Migrator controller then picks up this resource and executes the actual migration.
+
+When transitioning the Storage Version Migrator in-tree, we will exclusively move the Migrator controller as a component of KCM. The creation of the Migration resource will be deferred to the user, instead of being triggered automatically.
+
+#### Streaming List
+Currently, the Migrator controller uses the `chunked List` method to retrieve the list of resources from the API server and subsequently perform storage migrations as needed. However, chunked lists are [resource-intensive]((https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3157-watch-list/README.md#motivation)) and can lead to a significant overload on the API server, potentially resulting in it being terminated due to out-of-memory (OOM) issues. To address this concern, we have proposed the adoption of an Alpha feature introduced in Kubernetes _v1.27_, known as [`Streaming List`](https://kubernetes.io/docs/reference/using-api/api-concepts/#streaming-lists).
+
+When the Migrator controller is integrated in-tree, it will leverage the `Streaming List` approach to obtain and monitor resources while executing storage version migrations, as opposed to using the resource-intensive `chunked list` method. In cases where, for any reason, the client cannot establish a streaming watch connection with the API server, it will fall back to the standard `chunked list` method, retaining the older LIST/WATCH semantics.
+    
+- _Open Question_:
+    - Does `Streaming List` support inconsistent lists? Currently, with chunked lists, we receive inconsistent lists. We need to determine if we can achieve the same with Streaming List. Depending on the outcome, we may consider removing the [ContinueToken](https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/60dee538334c2366994c2323c0db5db8ab4d2838/pkg/apis/migration/v1alpha1/types.go#L63) from the API.
```

Comment:
@deads2k @jpbetz @wojtek-t @p0lyn0mial in an attempt to verify this, I [hacked in continue token support to streaming watch](https://github.com/kubernetes/kubernetes/commit/c760c43edb9871a8b8ab9696d41a4577d6a4b4a3).  While that implementation is obviously not what we would ship, would y'all be okay with the alpha SVM code simply using streaming watch and assuming it will work (i.e. for large resource sets it will fail to migrate if the watch cache or etcd are unable to provide a consistent list over the slow migration loop that SVM performs).  Assuming that streaming watch gives lexicographical ordering of events (it appears to, though I do not know if that is on purpose), it can support inconsistent continue tokens, at least if the caller has a way to bypass the watch cache.  @wojtek-t @p0lyn0mial is that something y'all would be willing to support as part of the streaming watch KEP?

Reply:
@p0lyn0mial

Reply:
Unfortunately, streaming doesn't support pagination (passing a continuation token) because the `watchcache` doesn't have built-in support for this mechanism.

I know that @wojtek-t  considered adding support ([as part of a new KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/365-paginated-lists#potential-future-extensions)) to the `watchcache` and even discussed it at a meeting, but I don’t know the current status.

As for using streaming in general, we plan to enable this feature for some component in this iteration. So, we really don't know if this mechanism works. We are about to find out.
I personally don't have anything against using the new mechanism for the alpha version of SVM.

Reply:
> Assuming that streaming watch gives lexicographical ordering of events

mhm, streaming as `give me a consistent list (snapshot) of data as a stream` gives you a random order. The main use is that you collect the data received from the server in some local store and only then operate on it (you don’t process individual events).

mhm, `watching` in general gives you a deterministic order (but not lexicographical) - the events are delivered in the order in which they occurred.

Reply:
Assuming you cannot use an informer (I haven't checked) please check out https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L330 to see how you could implement a fallback logic. 

At some point I wanted to also change the client-go's LIST method to follow streaming but I haven't looked into it yet.

Reply:
There are couple different things mentioned here that we're touching

> Assuming that streaming watch gives lexicographical ordering of events

This is only true for etcd implementation (because the list from etcd is returned in order).
This is not true for watchcache implementation as of now:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go#L120-L150

The fix would be trivial though - it just requires adding sorting there [and we should probably do that].


> it can support inconsistent continue tokens

My problem with that is that I'm not sure what that would effectively mean from the API perspective.
The way how watch-list currently works is that:
1. it lists all items [providing guarantees about freshness based on input parameters]
2. [if requests] returns a bookmark showing that "list" operation finished
3. continues to watch for items from that point [as part of the same request]

If we decide to support continuation token, (3) no longer makes sense to me [we started list from the middle of the collection, but watch is actually watching for the whole collection].
[I guess we would have to stop after point (2) then, but that still seems a bit sloppy...]

We also don't return continuation token when list-watch fails in the middle of listing phase [though that we can probably do].


>  at least if the caller has a way to bypass the watch cache.

If the caller bypasses watchcache, then the gain from watch-list is zero, because etcd implementation is doing a regular list from etcd. [We basically had need to ensure that API is supported if watch-cache is disabled, so we have both watchcache and etcd implementation, but the assumption is that watchcache implementation is the one that will generally be used and this is the one that actually provides real benefits].

That said, watchcache doesn't support pagination...


> would y'all be okay with the alpha SVM code simply using streaming watch and assuming it will work

That sounds reasonable to me for Alpha...
... but we need some solution for Beta.
And here my problem is that based on what I wrote above I'm not sure that supporting continuation for watch-list is the right path forward. And if not we would need a fallback to regular paginated list [and if so maybe we should just use paginated lists?]

Reply:
I get the impression that we mix two different use cases.

The first is that a client can collect all items in memory by asking for a stream of data that will result in a consistent snapshot.

The second is that a client doesn't need to store all items in memory and would like to work on individual items by asking for a consistent stream of data from the server. 

As of today the first use case is supported by the api streaming feature.
I assumed that the second would be supported when we implement pagination to the watchcache
But after the last message from @wojtek-t  I'm not sure.

Reply:
> If the caller bypasses watchcache, then the gain from watch-list is zero, because etcd implementation is doing a regular list from etcd

Ah yeah fair point.

> I'm not sure that supporting continuation for watch-list is the right path forward

I think there is a middle ground here - if we fix the ordering of the watch cache (so that we know what the next key is supposed to be), could it include an inconsistent continue token when it returns a `NewResourceExpired` error (with RV set to `-1`)?  Then when SVM makes the next call with the continue token set, we would bypass the watch cache and hit the regular etcd list (but for the subset of remaining items which should reduce the memory impact).

> And if not we would need a fallback to regular paginated list [and if so maybe we should just use paginated lists?]

@deads2k requested that we try to use watch list because it has a much lower memory impact on the API server.  Would we even need to support fallback?  We could feature gate SVM behind the watch list feature gate and require it to be enabled to use SVM, for example.

Reply:
> I think there is a middle ground here - if we fix the ordering of the watch cache (so that we know what the next key is supposed to be), could it include an inconsistent continue token when it returns a NewResourceExpired error (with RV set to -1)? 

That part sounds reasonable.

What I don't like is supporting `continueToken` for resuming watchlist, because the semantic of this would be strange:
- the init events are returned starting from continue token
- the following event chain is served for all objects

This may lead to very subtle issue if someone will not understand the consequences of doing that.

Reply:
FWIW - I opened https://github.com/kubernetes/kubernetes/pull/120897 to fix the ordering issue for watch-cache, so once that merges, the ordering problem is solved. But I'm still far from being convinced about supporting continue token...

Reply:
> What I don't like is supporting continueToken for resuming watchlist, because the semantic of this would be strange

Yeah I'm not crazy about doing something like this. It makes resumption subtly semantically different and I don't really like the idea of giving people foot guns.

Reply:
This is a tough problem, but I'd be disappointed if it prevented us from making forward progress in 1.29.

How about we stick with paginated list  for alpha and add a beta graduation criteria line item like "Resolve API server memory utilization problems with paginated list by enabling watch-list to handle large lists (either via a "inconsistent pagination token" mechanism or via some other approach)" ?

Reply:
I'm ok with defering to beta, but I see having a predictable and limited memory footprint as critical at scale.  The list OOMing problem, even with low hundreds of items, is a serious problem.

Reply:
I definitely agree that resolving it should block Beta.
But for Alpha, I'm with Joe - I suggest proceeding with paginated lists to unblock progress.

Can we mark it explicitly as <UNRESOLVED>